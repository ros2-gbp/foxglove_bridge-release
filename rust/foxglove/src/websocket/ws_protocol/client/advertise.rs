//! Client advertise message types.

use std::borrow::Cow;

use serde::{Deserialize, Serialize};

use crate::websocket::ws_protocol::schema::{self, Schema};
use crate::websocket::ws_protocol::JsonMessage;

/// Client advertise message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#client-advertise>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "advertise", rename_all = "camelCase")]
pub struct Advertise<'a> {
    /// Advertised channels.
    #[serde(borrow)]
    pub channels: Vec<Channel<'a>>,
}

impl<'a> Advertise<'a> {
    /// Creates a new advertise message
    pub fn new(channels: impl IntoIterator<Item = Channel<'a>>) -> Self {
        Self {
            channels: channels.into_iter().collect(),
        }
    }

    /// Returns an owned version of this message.
    pub fn into_owned(self) -> Advertise<'static> {
        Advertise {
            channels: self.channels.into_iter().map(|c| c.into_owned()).collect(),
        }
    }
}

impl JsonMessage for Advertise<'_> {}

/// Client channel advertisement.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Channel<'a> {
    /// Channel ID.
    pub id: u32,
    /// Topic name.
    #[serde(borrow)]
    pub topic: Cow<'a, str>,
    /// Message encoding.
    #[serde(borrow)]
    pub encoding: Cow<'a, str>,
    /// Schema name.
    #[serde(borrow)]
    pub schema_name: Cow<'a, str>,
    /// Schema encoding.
    #[serde(borrow, skip_serializing_if = "Option::is_none")]
    pub schema_encoding: Option<Cow<'a, str>>,
    /// Schema data.
    ///
    /// This is the protocol-encoded form. You can use the [`Channel::schema`] method to
    /// decode it.
    #[serde(borrow, skip_serializing_if = "Option::is_none")]
    pub schema: Option<Cow<'a, str>>,
}

impl<'a> Channel<'a> {
    /// Creates a new builder for a channel advertisement.
    #[must_use]
    pub fn builder(
        id: u32,
        topic: impl Into<Cow<'a, str>>,
        encoding: impl Into<Cow<'a, str>>,
    ) -> ChannelBuilder<'a> {
        ChannelBuilder {
            id,
            topic: topic.into(),
            encoding: encoding.into(),
            schema: None,
        }
    }

    /// Returns the decoded schema data.
    pub fn decode_schema(&self) -> Result<Vec<u8>, schema::DecodeError> {
        if let Some((schema_encoding, schema)) =
            self.schema_encoding.as_ref().zip(self.schema.as_ref())
        {
            schema::decode_schema_data(schema_encoding, schema)
        } else {
            Err(schema::DecodeError::MissingSchema)
        }
    }

    /// Returns an owned version of this channel.
    pub fn into_owned(self) -> Channel<'static> {
        Channel {
            id: self.id,
            topic: self.topic.into_owned().into(),
            encoding: self.encoding.into_owned().into(),
            schema_name: self.schema_name.into_owned().into(),
            schema_encoding: self.schema_encoding.map(|s| s.into_owned().into()),
            schema: self.schema.map(|s| s.into_owned().into()),
        }
    }
}

impl<'a> TryFrom<Channel<'a>> for Schema<'a> {
    type Error = schema::DecodeError;

    fn try_from(value: Channel<'a>) -> Result<Self, schema::DecodeError> {
        let schema = value.decode_schema()?;
        Ok(Schema::new(
            value.schema_name,
            value.schema_encoding.unwrap_or_default(),
            schema,
        ))
    }
}

/// Client channel advertisement builder.
pub struct ChannelBuilder<'a> {
    id: u32,
    topic: Cow<'a, str>,
    encoding: Cow<'a, str>,
    schema: Option<Schema<'a>>,
}

impl<'a> ChannelBuilder<'a> {
    /// Adds a schema to the channel advertisement.
    #[must_use]
    pub fn with_schema(mut self, schema: Schema<'a>) -> Self {
        self.schema = Some(schema);
        self
    }

    /// Constructs the channel advertisement.
    pub fn build(self) -> Result<Channel<'a>, schema::EncodeError> {
        match self.schema {
            None => {
                if schema::is_schema_required(&self.encoding) {
                    Err(schema::EncodeError::MissingSchema)
                } else {
                    Ok(Channel {
                        id: self.id,
                        topic: self.topic,
                        encoding: self.encoding,
                        schema_name: "".into(),
                        schema_encoding: None,
                        schema: None,
                    })
                }
            }
            Some(schema) => Ok(Channel {
                id: self.id,
                topic: self.topic,
                encoding: self.encoding,
                schema: Some(schema::encode_schema_data(&schema.encoding, schema.data)?),
                schema_name: schema.name,
                schema_encoding: Some(schema.encoding),
            }),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::websocket::ws_protocol::client::ClientMessage;

    use super::*;

    fn message() -> Advertise<'static> {
        Advertise::new([
            Channel::builder(10, "/t1", "json").build().unwrap(),
            Channel::builder(20, "/t2", "json")
                .with_schema(Schema::new(
                    "t2-schema",
                    "jsonschema",
                    br#"{"type": "object"}"#,
                ))
                .build()
                .unwrap(),
            Channel::builder(30, "/t3", "protobuf")
                .with_schema(Schema::new(
                    "t3-schema",
                    "protobuf",
                    &[0xde, 0xad, 0xbe, 0xef],
                ))
                .build()
                .unwrap(),
        ])
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = serde_json::to_string(&orig).unwrap();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::Advertise(orig));
    }
}
