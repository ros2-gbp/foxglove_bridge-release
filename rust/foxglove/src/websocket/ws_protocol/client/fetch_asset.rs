use serde::{Deserialize, Serialize};

use crate::websocket::ws_protocol::JsonMessage;

/// Fetch asset message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#fetch-asset>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "fetchAsset", rename_all = "camelCase")]
pub struct FetchAsset {
    /// Request ID.
    pub request_id: u32,
    /// Asset URI.
    pub uri: String,
}

impl FetchAsset {
    /// Creates a new fetch asset message.
    pub fn new(request_id: u32, uri: impl Into<String>) -> Self {
        Self {
            request_id,
            uri: uri.into(),
        }
    }
}

impl JsonMessage for FetchAsset {}

#[cfg(test)]
mod tests {
    use crate::websocket::ws_protocol::client::ClientMessage;

    use super::*;

    fn message() -> FetchAsset {
        FetchAsset::new(42, "package://foxglove/example.urdf")
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::FetchAsset(orig));
    }
}
