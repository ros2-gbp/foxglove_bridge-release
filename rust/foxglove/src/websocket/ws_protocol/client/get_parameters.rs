use serde::{Deserialize, Serialize};

use crate::websocket::ws_protocol::JsonMessage;

/// Get parameters message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#get-parameters>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "getParameters", rename_all = "camelCase")]
pub struct GetParameters {
    /// Parameter names.
    pub parameter_names: Vec<String>,
    /// Request ID.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
}

impl GetParameters {
    /// Creates a new get parameters message.
    pub fn new(parameter_names: impl IntoIterator<Item = impl Into<String>>) -> Self {
        Self {
            parameter_names: parameter_names.into_iter().map(|s| s.into()).collect(),
            id: None,
        }
    }

    /// Sets the request ID.
    #[must_use]
    pub fn with_id(mut self, id: impl Into<String>) -> Self {
        self.id = Some(id.into());
        self
    }
}

impl JsonMessage for GetParameters {}

#[cfg(test)]
mod tests {
    use crate::websocket::ws_protocol::client::ClientMessage;

    use super::*;

    fn message() -> GetParameters {
        GetParameters::new(["param1", "param2"])
    }

    fn message_with_id() -> GetParameters {
        message().with_id("my-id")
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_encode_with_id() {
        insta::assert_json_snapshot!(message_with_id());
    }

    fn test_roundtrip_inner(orig: GetParameters) {
        let buf = orig.to_string();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::GetParameters(orig));
    }

    #[test]
    fn test_roundtrip() {
        test_roundtrip_inner(message())
    }

    #[test]
    fn test_roundtrip_with_id() {
        test_roundtrip_inner(message_with_id())
    }
}
