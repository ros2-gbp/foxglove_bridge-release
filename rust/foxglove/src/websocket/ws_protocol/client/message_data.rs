use std::borrow::Cow;

use bytes::{Buf, BufMut};

use crate::websocket::ws_protocol::{BinaryMessage, ParseError};

use super::BinaryOpcode;

/// Client message data message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#client-message-data>
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MessageData<'a> {
    /// Channel ID.
    pub channel_id: u32,
    /// Message data.
    pub data: Cow<'a, [u8]>,
}

impl<'a> MessageData<'a> {
    /// Creates a new message data message.
    pub fn new(channel_id: u32, data: impl Into<Cow<'a, [u8]>>) -> Self {
        Self {
            channel_id,
            data: data.into(),
        }
    }

    /// Returns an owned version of this message.
    pub fn into_owned(self) -> MessageData<'static> {
        MessageData {
            channel_id: self.channel_id,
            data: Cow::Owned(self.data.into_owned()),
        }
    }
}

impl<'a> BinaryMessage<'a> for MessageData<'a> {
    fn parse_binary(mut data: &'a [u8]) -> Result<Self, ParseError> {
        if data.len() < 4 {
            return Err(ParseError::BufferTooShort);
        }
        let channel_id = data.get_u32_le();
        Ok(Self {
            channel_id,
            data: Cow::Borrowed(data),
        })
    }

    fn to_bytes(&self) -> Vec<u8> {
        let size = 1 + 4 + self.data.len();
        let mut buf = Vec::with_capacity(size);
        buf.put_u8(BinaryOpcode::MessageData as u8);
        buf.put_u32_le(self.channel_id);
        buf.put_slice(&self.data);
        buf
    }
}

#[cfg(test)]
mod tests {
    use crate::websocket::ws_protocol::client::ClientMessage;

    use super::*;

    fn message() -> MessageData<'static> {
        MessageData::new(30, br#"{"key": "value"}"#)
    }

    #[test]
    fn test_encode() {
        insta::assert_snapshot!(format!("{:#04x?}", message().to_bytes()));
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_bytes();
        let msg = ClientMessage::parse_binary(&buf).unwrap();
        assert_eq!(msg, ClientMessage::MessageData(orig));
    }
}
