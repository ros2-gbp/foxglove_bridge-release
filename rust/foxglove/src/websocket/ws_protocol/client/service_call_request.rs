use std::borrow::Cow;

use bytes::{Buf, BufMut};

use crate::websocket::ws_protocol::{BinaryMessage, ParseError};

use super::BinaryOpcode;

/// Service call request message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#service-call-request>
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ServiceCallRequest<'a> {
    /// Service ID.
    pub service_id: u32,
    /// Call ID.
    pub call_id: u32,
    /// Encoding.
    pub encoding: Cow<'a, str>,
    /// Payload.
    pub payload: Cow<'a, [u8]>,
}

impl ServiceCallRequest<'_> {
    /// Returns an owned version of this message.
    pub fn into_owned(self) -> ServiceCallRequest<'static> {
        ServiceCallRequest {
            service_id: self.service_id,
            call_id: self.call_id,
            encoding: Cow::Owned(self.encoding.into_owned()),
            payload: Cow::Owned(self.payload.into_owned()),
        }
    }
}

impl<'a> BinaryMessage<'a> for ServiceCallRequest<'a> {
    fn parse_binary(mut data: &'a [u8]) -> Result<Self, ParseError> {
        if data.remaining() < 4 + 4 + 4 {
            return Err(ParseError::BufferTooShort);
        }
        let service_id = data.get_u32_le();
        let call_id = data.get_u32_le();
        let encoding_length = data.get_u32_le() as usize;
        if data.remaining() < encoding_length {
            return Err(ParseError::BufferTooShort);
        }
        let encoding = Cow::Borrowed(std::str::from_utf8(&data[..encoding_length])?);
        data.advance(encoding_length);
        Ok(Self {
            service_id,
            call_id,
            encoding,
            payload: Cow::Borrowed(data),
        })
    }

    fn to_bytes(&self) -> Vec<u8> {
        let size = 1 + 4 + 4 + 4 + self.encoding.len() + self.payload.len();
        let mut buf = Vec::with_capacity(size);
        buf.put_u8(BinaryOpcode::ServiceCallRequest as u8);
        buf.put_u32_le(self.service_id);
        buf.put_u32_le(self.call_id);
        buf.put_u32_le(self.encoding.len() as u32);
        buf.put_slice(self.encoding.as_bytes());
        buf.put_slice(&self.payload);
        buf
    }
}

#[cfg(test)]
mod tests {
    use assert_matches::assert_matches;

    use crate::websocket::ws_protocol::client::ClientMessage;

    use super::*;

    fn message() -> ServiceCallRequest<'static> {
        ServiceCallRequest {
            service_id: 10,
            call_id: 12,
            encoding: "json".into(),
            payload: br#"{"key": "value"}"#.into(),
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_snapshot!(format!("{:#04x?}", message().to_bytes()));
    }

    #[test]
    fn test_parse() {
        assert_matches!(
            ServiceCallRequest::parse_binary(b""),
            Err(ParseError::BufferTooShort)
        );
        assert_matches!(
            ServiceCallRequest::parse_binary(&[0; 11]),
            Err(ParseError::BufferTooShort)
        );
        let mut buf = Vec::new();
        buf.put_u32_le(10);
        buf.put_u32_le(12);
        buf.put_u32_le(1);
        assert_matches!(
            ServiceCallRequest::parse_binary(&buf),
            Err(ParseError::BufferTooShort)
        );
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_bytes();
        let msg = ClientMessage::parse_binary(&buf).unwrap();
        assert_eq!(msg, ClientMessage::ServiceCallRequest(orig));
    }
}
