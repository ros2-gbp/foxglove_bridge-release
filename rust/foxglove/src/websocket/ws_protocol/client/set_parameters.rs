use serde::{Deserialize, Serialize};

use crate::websocket::ws_protocol::parameter::Parameter;
use crate::websocket::ws_protocol::JsonMessage;

/// Set parameters message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#set-parameters>
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "setParameters", rename_all = "camelCase")]
pub struct SetParameters {
    /// Parameters.
    pub parameters: Vec<Parameter>,
    /// Request ID.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
}

impl SetParameters {
    /// Creates a new set parameters message.
    pub fn new(parameters: impl IntoIterator<Item = Parameter>) -> Self {
        Self {
            parameters: parameters.into_iter().collect(),
            id: None,
        }
    }

    /// Sets the request ID.
    #[must_use]
    pub fn with_id(mut self, id: impl Into<String>) -> Self {
        self.id = Some(id.into());
        self
    }
}

impl JsonMessage for SetParameters {}

#[cfg(test)]
mod tests {
    use crate::websocket::ws_protocol::client::ClientMessage;
    use crate::websocket::ws_protocol::parameter::Parameter;

    use super::*;

    fn message() -> SetParameters {
        SetParameters::new([
            Parameter::empty("empty"),
            Parameter::integer("int", 123),
            Parameter::integer_array("int[]", [123, 456]),
            Parameter::float64("f64", 1.23),
            Parameter::float64_array("f64[]", [1.23, 4.56]),
            Parameter::string("str", "hello"),
            Parameter::byte_array("byte[]", &[0x10, 0x20, 0x30]),
            Parameter::bool("bool", true),
        ])
    }

    fn message_with_id() -> SetParameters {
        message().with_id("my-id")
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_encode_with_id() {
        insta::assert_json_snapshot!(message_with_id());
    }

    fn test_roundtrip_inner(orig: SetParameters) {
        let buf = orig.to_string();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::SetParameters(orig));
    }

    #[test]
    fn test_roundtrip() {
        test_roundtrip_inner(message())
    }

    #[test]
    fn test_roundtrip_with_id() {
        test_roundtrip_inner(message_with_id())
    }
}
