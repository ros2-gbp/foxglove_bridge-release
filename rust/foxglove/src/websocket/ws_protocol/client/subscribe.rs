//! Subscribe message types.

use serde::{Deserialize, Serialize};

use crate::websocket::ws_protocol::JsonMessage;

/// Subscribe message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#subscribe>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "subscribe", rename_all = "camelCase")]
pub struct Subscribe {
    /// Subscriptions.
    pub subscriptions: Vec<Subscription>,
}

impl Subscribe {
    /// Creates a new subscribe message.
    pub fn new(subscriptions: impl IntoIterator<Item = Subscription>) -> Self {
        Self {
            subscriptions: subscriptions.into_iter().collect(),
        }
    }
}

impl JsonMessage for Subscribe {}

/// A subscription for a [`Subscribe`] message.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Subscription {
    /// Subscription ID.
    pub id: u32,
    /// Channel ID.
    pub channel_id: u64,
}

impl Subscription {
    /// Creates a new subscription with the specified channel ID and subscription ID.
    pub fn new(id: u32, channel_id: u64) -> Self {
        Self { id, channel_id }
    }
}

#[cfg(test)]
mod tests {
    use crate::websocket::ws_protocol::client::ClientMessage;

    use super::*;

    fn message() -> Subscribe {
        Subscribe::new([Subscription::new(1, 10), Subscription::new(2, 20)])
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::Subscribe(orig));
    }
}
