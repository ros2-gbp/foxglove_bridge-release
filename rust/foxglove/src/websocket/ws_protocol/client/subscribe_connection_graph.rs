use serde::{Deserialize, Serialize};

use crate::websocket::ws_protocol::JsonMessage;

/// Subscribe connection graph message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#subscribe-connection-graph>
#[derive(Debug, Default, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(
    tag = "op",
    rename = "subscribeConnectionGraph",
    rename_all = "camelCase"
)]
pub struct SubscribeConnectionGraph {}

impl JsonMessage for SubscribeConnectionGraph {}

#[cfg(test)]
mod tests {
    use crate::websocket::ws_protocol::client::ClientMessage;

    use super::*;

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(SubscribeConnectionGraph {});
    }

    #[test]
    fn test_roundtrip() {
        let orig = SubscribeConnectionGraph {};
        let buf = orig.to_string();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::SubscribeConnectionGraph);
    }
}
