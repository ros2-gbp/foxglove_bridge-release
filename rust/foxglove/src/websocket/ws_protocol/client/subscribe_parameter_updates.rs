use serde::{Deserialize, Serialize};

use crate::websocket::ws_protocol::JsonMessage;

/// Subscribe parameter updates message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#subscribe-parameter-update>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(
    tag = "op",
    rename = "subscribeParameterUpdates",
    rename_all = "camelCase"
)]
pub struct SubscribeParameterUpdates {
    /// Parameter names.
    pub parameter_names: Vec<String>,
}

impl SubscribeParameterUpdates {
    /// Creates a new SubscribeParameterUpdates from an iterator of strings.
    pub fn new(names: impl IntoIterator<Item = impl Into<String>>) -> Self {
        Self {
            parameter_names: names.into_iter().map(Into::into).collect(),
        }
    }
}

impl JsonMessage for SubscribeParameterUpdates {}

#[cfg(test)]
mod tests {
    use crate::websocket::ws_protocol::client::ClientMessage;

    use super::*;

    fn message() -> SubscribeParameterUpdates {
        SubscribeParameterUpdates::new(["param1", "param2"])
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::SubscribeParameterUpdates(orig));
    }
}
