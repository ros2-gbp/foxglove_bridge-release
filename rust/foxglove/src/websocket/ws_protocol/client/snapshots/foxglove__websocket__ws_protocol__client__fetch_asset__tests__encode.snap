---
source: rust/foxglove/src/websocket/ws_protocol/client/fetch_asset.rs
expression: message()
---
{
  "op": "fetchAsset",
  "requestId": 42,
  "uri": "package://foxglove/example.urdf"
}
